### DISCUSSION TITLE
Any plans to support FFT?

### DISCUSSION THREAD

**Discussion by zh5j** [taichi-dev/taichi#8446] [Q&A] [answered]
FFT is a fundamental algorithm in digital signal processing, including image processing and RF/audio/video processing. 

As Taichi already support complex numbers and basic complex arithmetic, is there any plans to add something similar to cupy.fft.fft(a)? 

Or, is this type of function completely against Taichi's spirit in some aspect that I neglected, and I should implement my own version?

Any suggestions much appreciated!

**Comment by bobcao3** [comment, marked as answer]
> FFT is a fundamental algorithm in digital signal processing, including image processing and RF/audio/video processing. 
> 
> As Taichi already support complex numbers and basic complex arithmetic, is there any plans to add something similar to cupy.fft.fft(a)? 
> 
> Or, is this type of function completely against Taichi's spirit in some aspect that I neglected, and I should implement my own version?
> 
> Any suggestions much appreciated!

This is quite outside of the scope of taichi... We might be able to do some cupy / taichi interop so you can pass around the ndarrays. Supporting this outside of cuda (and cuFFT) is going to be significantly harder, maybe requires a taichi implement

**Comment by zh5j** [comment]
Understood, thanks for the info. Looking forward to the cupy interop then. I figured there has to be some way to pass GPU pointers around, something much like they are doing [here](https://github.com/vispy/vispy/pull/2429)... But anyway, thanks for the clarification, marking as answer.

**Comment by Mirebeau** [comment]
I was faced with this problem and used : 
- taichi.ndarray, which is compatible with pytorch GPU (and CPU) tensors
- torch.fft, which applies to pytorch GPU (and CPU) tensors

That being said, an interop with cupy would be greatly appreciated.
